core/src/main/java/org/elasticsearch/action/support/replication/TransportReplicationAction.java

```diff
@@ -235,17 +237,48 @@ public void messageReceived(final Request request, final TransportChannel channe
         }
 
         @Override
-        public void messageReceived(Request request, TransportChannel channel, Task task) throws Exception {
+        public void messageReceived(Request request, TransportChannel channel, Task task) {
             ReplicationTask replicationTask = (ReplicationTask) task;
-            boolean success = false;
-            PrimaryShardReference primaryShardReference = getPrimaryShardReference(request.shardId());
-            try {
-                if (primaryShardReference.isRelocated()) {
+
+            Callback<Throwable> onFailure = t -> {
```

Comment:
I would very much prefer an `AsyncPrimaryAction` (especially so that it implements AbstractRunnable, making failures simpler).